tracer/src/Datadog.Tracer.Native/clr_helpers.cpp

```diff
@@ -1018,7 +1018,318 @@ HRESULT HasAsyncStateMachineAttribute(const ComPtr<IMetaDataImport2>& metadataIm
+            continue;
+        }
+
+        const auto& candidateAssemblyImport =
+            metadata_interfaces.As<IMetaDataAssemblyImport>(IID_IMetaDataAssemblyImport);
+        const auto& candidateMetadataImport = metadata_interfaces.As<IMetaDataImport2>(IID_IMetaDataImport);
+
+        hr = candidateAssemblyImport->GetAssemblyFromScope(&candidateAssembly);
+        if (FAILED(hr))
+        {
+            Logger::Warn("[ResolveTypeInternal] GetAssemblyFromScope has failed with: ", shared::WSTRING(refTypeName.data()));
+            continue;
+        }
+
+        const auto& assemblyMetadata = GetAssemblyImportMetadata(candidateAssemblyImport);
+
+        if (assemblyMetadata.name == shared::EmptyWStr)
+        {
+            Logger::Warn("[ResolveTypeInternal] GetAssemblyImportMetadata has failed with: ", shared::WSTRING(refTypeName.data()));
+            continue;
+        }
+
+        const auto& candidateResolutionScopeName = assemblyMetadata.name;
+
+        if (resolutionScopeName == candidateResolutionScopeName)
+        {
+            hr = candidateMetadataImport->FindTypeDefByName(refTypeName.data(), parentToken, &resolvedTypeDefToken);
+            if (hr == S_OK)
+            {
+                resolvedTypeDefMetadataImport = candidateMetadataImport;
+                isModuleFound = true;
+                break;
+            }
+            else
+            {
+                // look for exported type with the same name of the given TypeRef
+                mdExportedType exportedType;
+                hr = candidateAssemblyImport->FindExportedTypeByName(refTypeName.data(), parentToken, &exportedType);
+                if (FAILED(hr))
+                {
+                    Logger::Warn("[ResolveTypeInternal] FindExportedTypeByName has failed with: ", shared::WSTRING(refTypeName.data()));
+                    continue;
+                }
+
+                if (hr == S_OK)
+                {
+                    WCHAR szName[kNameMaxSize];
+                    ULONG pchName;
+                    mdToken exportedTypeContainerToken; // will hold mdAssemblyRef / mdExportedType  / mdFile
+                    mdTypeDef exportedTypeTypeDef;
+                    DWORD exportedTypeFlags;
+
+                    hr = candidateAssemblyImport->GetExportedTypeProps(exportedType, szName, kNameMaxSize, &pchName,
+                                                                       &exportedTypeContainerToken,
+                                                                       &exportedTypeTypeDef, &exportedTypeFlags);
+                    if (FAILED(hr))
+                    {
+                        Logger::Warn("[ResolveTypeInternal] GetExportedTypeProps [1] has failed with: ", shared::WSTRING(refTypeName.data()));
+                        continue;
+                    }
+
+                    int tryoutsCount = 1000; // To avoid falling into an infinite loop
```

Comment:
Not sure it matters that much, but 1000 seems a bit steep. Can you think of an example where 100 won't be sufficient?

Reply:
It's pretty much arbitrary, we can have it at 100 but I think it's not that important. It's only there for a safe guard. Nothing more nothing less. It should be big enough to avoid false-positives, for sure. But considering the rarity of ever reaching it I think it's negligible.

Reply:
Fair enough, resolving